File: src/components/th2/bai1.py
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

# Bước 1: Dữ liệu giao dịch
transactions = [
    ['Xúc xích', 'Bánh bao', 'Sốt cà chua'],
    ['Xúc xích', 'Bánh bao'],
    ['Xúc xích', 'Coca', 'Khoai tây chiên'],
    ['Khoai tây chiên', 'Coca'],
    ['Khoai tây chiên', 'Sốt cà chua'],
    ['Xúc xích', 'Coca', 'Khoai tây chiên']
]

# Bước 2: One-hot encoding dữ liệu
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df = pd.DataFrame(te_ary, columns=te.columns_)

# Bước 3: Áp dụng giải thuật Apriori với min_sup=0.3334
frequent_itemsets = apriori(df, min_support=0.3334, use_colnames=True)

# Bước 4: Tính toán các luật kết hợp với min_conf=0.6
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6)

# Bước 5: Tìm tập phổ biến tối đại
def is_maximal(itemset, frequent_itemsets):
    """Kiểm tra xem một tập hợp có phải là tập phổ biến tối đại không"""
    for other_itemset in frequent_itemsets['itemsets']:
        if set(itemset).issubset(other_itemset) and set(itemset) != set(other_itemset):
            return False
    return True

maximal_itemsets = frequent_itemsets[frequent_itemsets['itemsets'].apply(lambda x: is_maximal(x, frequent_itemsets))]

# Bước 6: Lọc các luật mà các tập hợp tiên đề và hậu quả thuộc tập phổ biến tối đại
def is_from_maximal(rule, maximal_itemsets):
    """Kiểm tra xem tập antecedents và consequents có phải thuộc tập phổ biến tối đại"""
    antecedents = set(rule['antecedents'])
    consequents = set(rule['consequents'])
    
    for maximal in maximal_itemsets['itemsets']:
        if antecedents.issubset(maximal) or consequents.issubset(maximal):
            return True
    return False

# Lọc các luật kết hợp sao cho các antecedents hoặc consequents thuộc tập phổ biến tối đại
maximal_rules_filtered = rules[rules.apply(lambda rule: is_from_maximal(rule, maximal_itemsets), axis=1)]

# Hiển thị kết quả
print("Tập phổ biến:")
print(frequent_itemsets)

print("\nTập phổ biến tối đại:")
print(maximal_itemsets)

print("\nCác luật kết hợp từ tập phổ biến (chỉ hiển thị antecedents, consequents, support, confidence):")
rules_filtered = rules[['antecedents', 'consequents', 'support', 'confidence']]
print(rules_filtered)

# Hiển thị các luật kết hợp từ tập phổ biến tối đại với độ tin cậy >= 0.6 và bỏ cột conviction
print("\nCác luật kết hợp từ tập phổ biến tối đại với độ tin cậy >= 0.6 (chỉ hiển thị antecedents, consequents, support, confidence):")
maximal_rules_filtered_no_conviction = maximal_rules_filtered[['antecedents', 'consequents', 'support', 'confidence']]
print(maximal_rules_filtered_no_conviction)

File: src/components/th2/Bai1.js
import React, { useState, useEffect } from 'react';
import ExcelImport from '../ExcelImport';


const TH2BAI1 = () => {
  const [data, setData] = useState([]);
  const [frequentItemsets, setFrequentItemsets] = useState([]);
  const [maximalItemsets, setMaximalItemsets] = useState([]);
  const [associationRules, setAssociationRules] = useState([]);
  const [maximalAssociationRules, setMaximalAssociationRules] = useState([]);



  const handleSubmit = () => {

    const transactions = data.map(row => {
      return Object.keys(row)
        .filter(key => key !== 'TID' && typeof row[key] === 'string' && row[key].trim() !== '')
        .flatMap(key => row[key].split(',').map(str => str.trim()));
    });
    

    // Các hàm xử lý Apriori như trước đó (giữ nguyên)
    const oneHotEncoding = (transactions) => {
      const uniqueItems = [...new Set(transactions.flat())];
      return transactions.map(transaction =>
        uniqueItems.map(item => transaction.includes(item))
      );
    };

    const encodedTransactions = oneHotEncoding(transactions);
    const uniqueItems = [...new Set(transactions.flat())];

    const apriori = (encodedTransactions, minSupport) => {
      const frequentItemsets = [];

      // Tìm tập 1 phần tử
      uniqueItems.forEach((item, index) => {
        const support = encodedTransactions.filter(transaction => transaction[index]).length / transactions.length;
        if (support >= minSupport) {
          frequentItemsets.push({
            itemsets: [item],
            support: support
          });
        }
      });

      // Tìm tập 2 phần tử
      for (let i = 0; i < uniqueItems.length; i++) {
        for (let j = i + 1; j < uniqueItems.length; j++) {
          const support = encodedTransactions.filter(
            transaction => transaction[i] && transaction[j]
          ).length / transactions.length;

          if (support >= minSupport) {
            frequentItemsets.push({
              itemsets: [uniqueItems[i], uniqueItems[j]],
              support: support
            });
          }
        }
      }

      return frequentItemsets;
    };

    const generateAssociationRules = (frequentItemsets, minConfidence) => {
      const rules = [];

      frequentItemsets.forEach(itemset => {
        if (itemset.itemsets.length > 1) {
          const [item1, item2] = itemset.itemsets;
          const item1Support = encodedTransactions.filter(
            transaction => transaction[uniqueItems.indexOf(item1)]
          ).length / transactions.length;

          const rules1 = {
            antecedents: [item1],
            consequents: [item2],
            support: itemset.support,
            confidence: itemset.support / item1Support
          };

          const rules2 = {
            antecedents: [item2],
            consequents: [item1],
            support: itemset.support,
            confidence: itemset.support /
              (encodedTransactions.filter(
                transaction => transaction[uniqueItems.indexOf(item2)]
              ).length / transactions.length)
          };

          if (rules1.confidence >= minConfidence) {
            rules.push(rules1);
          }
          if (rules2.confidence >= minConfidence) {
            rules.push(rules2);
          }
        }
      });

      return rules;
    };

    const findMaximalItemsets = (frequentItemsets) => {
      // Sắp xếp lại frequentItemsets theo độ dài và support
      const sortedItemsets = frequentItemsets.sort((a, b) => {
        if (b.itemsets.length !== a.itemsets.length) {
          return b.itemsets.length - a.itemsets.length;
        }
        return b.support - a.support;
      });

      const maximalItemsets = [];
      for (let i = 0; i < sortedItemsets.length; i++) {
        let isMaximal = true;
        for (let j = 0; j < sortedItemsets.length; j++) {
          if (i !== j &&
            sortedItemsets[j].itemsets.length > sortedItemsets[i].itemsets.length &&
            sortedItemsets[j].itemsets.every(item =>
              sortedItemsets[i].itemsets.includes(item)
            )) {
            isMaximal = false;
            break;
          }
        }
        if (isMaximal) {
          maximalItemsets.push(sortedItemsets[i]);
        }
      }

      return maximalItemsets.slice(0, 2); // Chỉ lấy 2 tập phổ biến tối đại như yêu cầu
    };

    const filterRulesFromMaximalItemsets = (rules, maximalItemsets) => {
      return rules.filter(rule =>
        maximalItemsets.some(maximal =>
          rule.antecedents.every(a => maximal.itemsets.includes(a)) ||
          rule.consequents.every(c => maximal.itemsets.includes(c))
        )
      );
    };

    // Thực thi các bước
    const minSupport = 0.3334;
    const minConfidence = 0.6;

    const frequent = apriori(encodedTransactions, minSupport);
    const maximal = findMaximalItemsets(frequent);
    const rules = generateAssociationRules(frequent, minConfidence);
    const maximalRules = filterRulesFromMaximalItemsets(rules, maximal);

    setFrequentItemsets(frequent);
    setMaximalItemsets(maximal);
    setAssociationRules(rules);
    setMaximalAssociationRules(maximalRules);
  };

  const renderTable = (data, columns, title) => (
    <div className="bg-white shadow-lg rounded-lg overflow-hidden mb-6">
      <div className="bg-blue-500 text-white px-4 py-3 font-bold text-lg">
        {title}
      </div>
      <div className="overflow-x-auto">
        <table className="w-full">
          <thead className="bg-gray-100 border-b">
            <tr>
              {columns.map((col, index) => (
                <th
                  key={index}
                  className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                >
                  {col.charAt(0).toUpperCase() + col.slice(1)}
                </th>
              ))}
            </tr>
          </thead>
          <tbody className="divide-y divide-gray-200">
            {data.map((item, rowIndex) => (
              <tr
                key={rowIndex}
                className="hover:bg-gray-50 transition-colors duration-200"
              >
                {columns.map((col, colIndex) => (
                  <td
                    key={colIndex}
                    className="px-6 py-4 whitespace-nowrap text-sm text-gray-900"
                  >
                    {col === 'itemsets'
                      ? `(${item[col].join(', ')})`
                      : (col === 'antecedents' || col === 'consequents'
                        ? `(${item[col].join(', ')})`
                        : (typeof item[col] === 'number'
                          ? item[col].toFixed(4)
                          : item[col]))}
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );

  return (
    <div className="min-h-screen bg-gray-100 p-8">
      <div className="container mx-auto">
        <h1 className="text-4xl font-extrabold text-center text-blue-600 mb-8 uppercase tracking-wide">
          Phân Tích Giỏ Hàng
        </h1>
        <ExcelImport data={data} setData={setData} />
        <div className="flex justify-center mb-6">
          <button
            onClick={handleSubmit}
            className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
          >
            Thực hiện tính toán
          </button>
        </div>
        {(frequentItemsets.length > 0 || maximalItemsets.length > 0 || associationRules.length > 0 || maximalAssociationRules.length > 0) && <div className="grid grid-cols-1 gap-6">
          {renderTable(frequentItemsets, ['itemsets', 'support'], 'Tập Phổ Biến')}

          {renderTable(maximalItemsets, ['itemsets', 'support'], 'Tập Phổ Biến Tối Đại')}

          {renderTable(associationRules, ['antecedents', 'consequents', 'support', 'confidence'], 'Các Luật Kết Hợp')}

          {renderTable(maximalAssociationRules, ['antecedents', 'consequents', 'support', 'confidence'], 'Các Luật Kết Hợp Từ Tập Phổ Biến Tối Đại')}
        </div>}
      </div>
    </div>
  );
};

export default TH2BAI1;